jsk_pcl_ros/src/color_based_region_growing_segmentation_nodelet.cpp

```diff
@@ -0,0 +1,91 @@
+#include <pcl_conversions/pcl_conversions.h>
+
+#include "jsk_pcl_ros/color_based_region_growing_segmentation.h"
+
+#include "jsk_recognition_utils/pcl_conversion_util.h"
+#include "jsk_recognition_msgs/ClusterPointIndices.h"
+
+
+namespace jsk_pcl_ros
+{
+
+  void ColorBasedRegionGrowingSegmentation::onInit()
+  {
+    ConnectionBasedNodelet::onInit();
+    srv_ = boost::make_shared <dynamic_reconfigure::Server<Config> > (*pnh_);
+    dynamic_reconfigure::Server<Config>::CallbackType f=
+      boost::bind (&ColorBasedRegionGrowingSegmentation::configCallback, this, _1, _2);
+    srv_->setCallback (f);
+    pub_ = advertise<jsk_recognition_msgs::ClusterPointIndices>(*pnh_, "output", 1);
+    onInitPostProcess();
+  }
+
+  void ColorBasedRegionGrowingSegmentation::subscribe()
+  {
+    sub_ = pnh_->subscribe("input", 1, &ColorBasedRegionGrowingSegmentation::segment, this);
+  }
+
+  void ColorBasedRegionGrowingSegmentation::unsubscribe()
+  {
+    sub_.shutdown();
+  }
+
+  void ColorBasedRegionGrowingSegmentation::configCallback(Config &config, uint32_t level)
+  {
+    boost::mutex::scoped_lock lock(mutex_);
+
+    if (distance_threshould_ != config.distance_threshould) {
+      distance_threshould_ = config.distance_threshould;
+    }
+    if (point_color_threshould_ != config.point_color_threshould) {
+      point_color_threshould_ = config.point_color_threshould;
+    }
+    if (region_color_threshould_ != config.region_color_threshould) {
+      region_color_threshould_ = config.region_color_threshould;
+    }
+    if (min_cluster_size_ != config.min_cluster_size) {
+      min_cluster_size_ = config.min_cluster_size;
+    }
+  }
+
+  void ColorBasedRegionGrowingSegmentation::segment(const sensor_msgs::PointCloud2::ConstPtr& msg)
+  {
+    boost::mutex::scoped_lock lock(mutex_);
+    pcl::search::Search <pcl::PointXYZRGB>::Ptr tree = boost::shared_ptr<pcl::search::Search<pcl::PointXYZRGB> > (new pcl::search::KdTree<pcl::PointXYZRGB>);
```

Comment:
pcl::search::Search <pcl::PointXYZRGB>::Ptr is better than boost::shared_ptr<pcl::search::Search<pcl::PointXYZRGB>>